diff --git a/blink/renderer/core/dom/document_timing.cc b/blink/renderer/core/dom/document_timing.cc
index af75c2da2..dc989394b 100644
--- a/blink/renderer/core/dom/document_timing.cc
+++ b/blink/renderer/core/dom/document_timing.cc
@@ -35,11 +35,10 @@ void DocumentTiming::MarkDomLoading() {
 #if BUILDFLAG(IS_OHOS)
   TRACE_EVENT1("navigation", "PAGE_LOAD_TIME",
                "domLoading", dom_loading_);
-#else
+#endif
   TRACE_EVENT_MARK_WITH_TIMESTAMP1("blink.user_timing,rail", "domLoading",
                                    dom_loading_, "frame",
                                    GetFrameIdForTracing(GetFrame()));
-#endif
   NotifyDocumentTimingChanged();
 }
 
@@ -48,11 +47,10 @@ void DocumentTiming::MarkDomInteractive() {
 #if BUILDFLAG(IS_OHOS)
   TRACE_EVENT1("navigation", "PAGE_LOAD_TIME",
                "domInteractive", dom_interactive_);
-#else
+#endif
   TRACE_EVENT_MARK_WITH_TIMESTAMP1("blink.user_timing,rail", "domInteractive",
                                    dom_interactive_, "frame",
                                    GetFrameIdForTracing(GetFrame()));
-#endif
   NotifyDocumentTimingChanged();
 }
 
@@ -61,12 +59,11 @@ void DocumentTiming::MarkDomContentLoadedEventStart() {
 #if BUILDFLAG(IS_OHOS)
   TRACE_EVENT1("navigation", "PAGE_LOAD_TIME",
                "domContentLoadedEventStart", dom_content_loaded_event_start_);
-#else
+#endif
   TRACE_EVENT_MARK_WITH_TIMESTAMP1("blink.user_timing,rail",
                                    "domContentLoadedEventStart",
                                    dom_content_loaded_event_start_, "frame",
                                    GetFrameIdForTracing(GetFrame()));
-#endif
   NotifyDocumentTimingChanged();
 }
 
@@ -75,11 +72,10 @@ void DocumentTiming::MarkDomContentLoadedEventEnd() {
 #if BUILDFLAG(IS_OHOS)
   TRACE_EVENT1("navigation", "PAGE_LOAD_TIME",
                "domContentLoadedEventEnd", dom_content_loaded_event_end_);
-#else
+#endif
   TRACE_EVENT_MARK_WITH_TIMESTAMP1(
       "blink.user_timing,rail", "domContentLoadedEventEnd",
       dom_content_loaded_event_end_, "frame", GetFrameIdForTracing(GetFrame()));
-#endif
   InteractiveDetector* interactive_detector(
       InteractiveDetector::From(*document_));
   if (interactive_detector) {
diff --git a/blink/renderer/core/loader/document_load_timing.cc b/blink/renderer/core/loader/document_load_timing.cc
index 60cfa37b5..747bbac34 100644
--- a/blink/renderer/core/loader/document_load_timing.cc
+++ b/blink/renderer/core/loader/document_load_timing.cc
@@ -140,14 +140,12 @@ void DocumentLoadTiming::SetNavigationStart(base::TimeTicks navigation_start) {
   // been set yet in order to have a valid reference time in both units.
   EnsureReferenceTimesSet();
   navigation_start_ = navigation_start;
-#if !BUILDFLAG(IS_OHOS)
   TRACE_EVENT_MARK_WITH_TIMESTAMP2(
       "blink.user_timing", "navigationStart", navigation_start_, "frame",
       GetFrameIdForTracing(GetFrame()), "data",
       [&](perfetto::TracedValue context) {
         WriteNavigationStartDataIntoTracedValue(std::move(context));
       });
-#endif
 
   // The reference times are adjusted based on the embedder's navigationStart.
   DCHECK(!reference_monotonic_time_.is_null());
@@ -203,21 +201,17 @@ void DocumentLoadTiming::AddRedirect(const KURL& redirecting_url,
 
 void DocumentLoadTiming::SetRedirectStart(base::TimeTicks redirect_start) {
   redirect_start_ = redirect_start;
-#if !BUILDFLAG(IS_OHOS)
   TRACE_EVENT_MARK_WITH_TIMESTAMP1("blink.user_timing", "redirectStart",
                                    redirect_start_, "frame",
                                    GetFrameIdForTracing(GetFrame()));
-#endif
   NotifyDocumentTimingChanged();
 }
 
 void DocumentLoadTiming::SetRedirectEnd(base::TimeTicks redirect_end) {
   redirect_end_ = redirect_end;
-#if !BUILDFLAG(IS_OHOS)
   TRACE_EVENT_MARK_WITH_TIMESTAMP1("blink.user_timing", "redirectEnd",
                                    redirect_end_, "frame",
                                    GetFrameIdForTracing(GetFrame()));
-#endif
   NotifyDocumentTimingChanged();
 }
 
@@ -243,21 +237,17 @@ void DocumentLoadTiming::MarkFetchStart() {
 
 void DocumentLoadTiming::SetFetchStart(base::TimeTicks fetch_start) {
   fetch_start_ = fetch_start;
-#if !BUILDFLAG(IS_OHOS)
   TRACE_EVENT_MARK_WITH_TIMESTAMP1("blink.user_timing", "fetchStart",
                                    fetch_start_, "frame",
                                    GetFrameIdForTracing(GetFrame()));
-#endif
   NotifyDocumentTimingChanged();
 }
 
 void DocumentLoadTiming::SetResponseEnd(base::TimeTicks response_end) {
   response_end_ = response_end;
-#if !BUILDFLAG(IS_OHOS)
   TRACE_EVENT_MARK_WITH_TIMESTAMP1("blink.user_timing", "responseEnd",
                                    response_end_, "frame",
                                    GetFrameIdForTracing(GetFrame()));
-#endif
   NotifyDocumentTimingChanged();
 }
 
@@ -266,11 +256,10 @@ void DocumentLoadTiming::MarkLoadEventStart() {
 #if BUILDFLAG(IS_OHOS)
   TRACE_EVENT1("navigation", "PAGE_LOAD_TIME",
                "loadEventStart", load_event_start_);
-#else
+#endif
   TRACE_EVENT_MARK_WITH_TIMESTAMP1("blink.user_timing", "loadEventStart",
                                    load_event_start_, "frame",
                                    GetFrameIdForTracing(GetFrame()));
-#endif
   NotifyDocumentTimingChanged();
 }
 
@@ -279,11 +268,10 @@ void DocumentLoadTiming::MarkLoadEventEnd() {
 #if BUILDFLAG(IS_OHOS)
   TRACE_EVENT1("navigation", "PAGE_LOAD_TIME",
                "loadEventEnd", load_event_end_);
-#else
+#endif
   TRACE_EVENT_MARK_WITH_TIMESTAMP1("blink.user_timing", "loadEventEnd",
                                    load_event_end_, "frame",
                                    GetFrameIdForTracing(GetFrame()));
-#endif
   NotifyDocumentTimingChanged();
 }
 
